### PR TITLE
Pixel cluster shape: new calibration for Phase1 and Phase2

### PR DESCRIPTION
**Pull request by VinInn**
This PR is composed of several components
1) A new version of the PixelClusterShapeExtractor that tries to deal with spit/merged clusters at best
3) A modified version of the ClusterShapeHitFilter that makes provision to use a different file for the first layer of the Barrel as it is the only one affected by long clusters and in 2017 data affected by inefficiency
4) A new version of the calibration files for Phase1 and Phase2
5) a rearrangement of DQM to monitor the ClusterShape
6) a python notebook to analyze and compare calibration files
7) new configuration files to perform the "extraction" 
8) updated documentation

The main reason of a new calibration is the changes in geometry (in particular for the forward pixel)
that happened in 910_pre3.
The new calibrations are based on 90K MinBias events from recent RelVal

MVA
for phase1
http://innocent.home.cern.ch/innocent/RelVal/pu35_candpcsf/
for phase2
http://innocent.home.cern.ch/innocent/RelVal/newPCSF2/

for phase1 there is also a comparison of using 
1. the calibration obtained considering all layers
2. the calibration obtained considering all layers but BPIX1
3. and using the file " obtained considering all layers but BPIX1" for all layers but BPIX1 (for BPIX1 the file obtained considering all layers is used)
http://innocent.home.cern.ch/innocent/RelVal/pu35_newpcsf/plots_highPurity/effandfake1.pdf

for Phase2 there is no difference.
for Phase1 there is no obvious improvements but more seeds in the forward region
(this is observable also in DQM for both data and mc) i.e. higher efficiency for quadruplets at eta~2
there is also an increase in fake rate in the barrel (the effect on late iterations are more difficult to establish as they are affected by the changes in the previous one)
Still this is the best calibration we can produce and the previous one was wrong as not covering properly the phasespace in the forward region (this calibration is the solution of an inverse problem, so it depends on the sample on which is trained)

It shall be noted that MVA (in particular for low-pt) is under retraining and shall be retrained once this PR is merged.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
On 7/19/17 7:27 AM, Vincenzo Innocente wrote:
> for phase1 there is also a comparison of using
> 
>  1. the calibration obtained considering all layers
>  2. the calibration obtained considering all layers but BPIX1
>  3. and using the file " obtained considering all layers but BPIX1" for
>     all layers but BPIX1 (for BPIX1 the file
>     obtained considering all layers is used)
>     http://innocent.home.cern.ch/innocent/RelVal/pu35_newpcsf/plots_highPurity/effandfake1.pdf

please clarify which of the 4 histograms is the baseline (which 93X is it?)
and which one is the proposed version with this PR.
I couldn't understand it from the text in this PR description.

Thank you.

**Comment by VinInn** [comment]
@slava77 look at http://innocent.home.cern.ch/innocent/RelVal/pu35_candpcsf/
he current one is blue and proposed is red
in any case in http://innocent.home.cern.ch/innocent/RelVal/pu35_newpcsf the current one is 
blue and proposed is red as well

**Comment by VinInn** [comment]
everything is with IBs of the last 10 days relval is pre2

**Comment by slava77** [comment]
10024 step2:
``` MissingParameter: Parameter 'PixelShapeFileL1' not found. ```
seems related

**Comment by VinInn** [comment]
HLT? I run the matrix.... will have a look

**Comment by VinInn** [comment]
could somebody remind me which file need to be edited to fix HLT config?

**Comment by perrotta** [comment]
Temporary custimizations to be applied before the updates are parsed  
in ConfDb are dealt with by
https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/customizeHLTforCMSSW.py

Vincenzo Innocente <notifications@github.com> ha scritto:

**Comment by VinInn** [comment]
ok, how one deals with eras?

**Comment by perrotta** [comment]
I would do it there... but probably @Martin-Grunewald can suggest you which is the best and most up to date procedure for it.

**Comment by VinInn** [comment]
I see at the end of the file something done for pixel calib...
```
if (menuType == "GRun2016"):
```
will add in there

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by Martin-Grunewald** [comment]
@VinInn 
Please rebase to get rid of spurious commits/PR merges.

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by dmitrijus** [comment]
+1

File: RecoPixelVertexing/PixelLowPtUtilities/data/READE.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+### Cluster shape calibration data files
+
+1. stripShape.par : original strip cluster shape (not used?)
+2. pixelShapePhase0.par  : original phase0 pixel calibration
+3. pixelShapePhase1_all.par  : pixel cluster shape for Phase1 produced with 90K MinBias 930_pre1 relval at 99.5%
+4. pixelShapePhase1_noL1.par : same w/o considering BPIX1
+4. ITShapePhase2_all.par : Inner Tracker pixel cluster shape for Phase2 produced with 90K MinBias 930_pre2 relval at 99.9%  
+4. ITShapePhase2_noL1.par : same w/o considering BPIX1
```
Comment (slava77) [inline RecoPixelVertexing/PixelLowPtUtilities/data/READE.md]: 4. 5. 6. instead of 4. 4. 4.?

**Comment by slava77** [comment]
the functionality derived from RecoPixelVertexing/PixelLowPtUtilities/data/pixelShape\* better appear in DB/GT soon.
If I understand correctly, it will be the best way to cover data processing in 2017 if we were to catch up with the differences in case1, case2, case 12' etc

Does it make sense?
If so, is it in the plan of work somewhere between TRK POG and DPG?

File: RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc [hunk 2]
```diff
@@ -0,0 +1,407 @@
+// VI January 2012: needs to be migrated to use cluster directly
```
Comment (slava77) [inline RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc]: please move this to /test to remove dependency on Sim for the lib used in production

File: RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+#include <vector>
+#include <fstream>
+
+#include<mutex>
+
+using namespace std;
+
+// pixel
+#define exMax 10
+#define eyMax 15
```
Comment (slava77) [inline RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc]: CMS coding rules say not to use macros for constants
.. will not be enforced in /test

File: RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+          // Fill map
+          if(simHitMap.count(key) == 0)
+              { simHitMap[key] = {&recHit,simHit,ss}; }
+          else if( recHit.cluster()->size() >
+                   simHitMap[key].rhit->cluster()->size())
+                   simHitMap[key] = {&recHit,simHit,std::max(ss,simHitMap[key].size)};
+  }
+  for (auto const & elem : simHitMap)  {
+   /* irrelevant
+   auto const rh = *elem.second.rhit;
```
Comment (slava77) [inline RecoPixelVertexing/PixelLowPtUtilities/plugins/PixelClusterShapeExtractor.cc]: comments.
again, OK if in /test

File: RecoPixelVertexing/PixelLowPtUtilities/test/pcsfVerify.ipynb [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
...Z\np2XLlutzzjlnJUk2bVqb887bu4+NG9flggtWHnR/c7Fhw9pceOHBawOAPquqtNbm9I3joLe63pxk\n36e4m5Lct5ueTHLqHAq8W1Wd0k0fn+SHk3xuLgUDAAAciUGHvX0qySOTXJPk/UleVVXnJfmvcxz2\ndo8kb6m940MWJfn71tpHjrB2AACAgQ0afn43yYnd9OokJyV5cpIvd8sG0lq7OsmDj6xUAACAIzfo\nQ06vmza9LclvDLUqAACAeTboNT+pqqVV9dNV9cKqWt61fXdV3WWoFQIAAMyDQZ/zc98kH+6Gvi1P\n8o4kt3Y9QMuTPHP4pQIAABy5QXt+XpPkkiR3T3L7tPb3JnnMkGoDAACYN4Pe8OCCJA9vre3e9yC/\nzo1J7jmc0gAAAObPwNf8JFk8S9sZSW6bx3oAAACGYtDwc8mMW1q3qjo5ycu65/4AAAAc0+bynJ9L\nq+pLSZYm+cck901yc5KfHXKNAAAAR23Q5/ysr6oHJXlq95DSRUlen+StrbXbB9gFAADAWA3a85Mu\n5Pxt9wMAALCgDBx+quq0JI9MctrMa4Vaa381lOoAAADmyaAPOX1q1+OzKMnGJG3a4pZE+AEAAI5p\ng/b8/HGSP03yB621qSHXBAAAMO8GvdX1KUneLPgAAAAL1aDh5x+SPGHItQAAAAzNoMPefjvJe6vq\ncUmuTrJr+sLW2h8MpzwAAID5MWj4+bUkP5zklu7hpjNveCD8AAAAx7RBw89Lkjy3tfbqIdcDAAAw\nFINe8zOR5L1DrgUAAGBoBg0/FyX5hSHXAgAAMDSDDns7Ickzq+rxSa6a5YYHzx5OeQAAAPNj0PCz\nMslnu+n7zVjWZlkfAADgmDJQ+GmtPWb4pQAAAAzPoNf8AAAALGjCDwAA0AvCDwAA0AvCDwAA0AvC\nDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA\n0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvC\nDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AvCDwAA0AsjDT9V\nde+q+mhVfaGqrq6qZ4/y+AAAQH9
```
Comment (slava77) [inline RecoPixelVertexing/PixelLowPtUtilities/test/pcsfVerify.ipynb]: is there a simpler way to include figures, compared to raw dump inlined as a string?

Comment (VinInn) [inline RecoPixelVertexing/PixelLowPtUtilities/test/pcsfVerify.ipynb]: this is standard jupyter: not under my control. very common in the outer world

**Comment by slava77** [comment]
On 7/21/17 9:44 PM, Vincenzo Innocente wrote:
> this is standard jupyter: not under my control. very common in the outer
> world
> 

good thing it's a small image.

So, no other syntax to include a .png file by name other than this
inline option?

File: RecoPixelVertexing/PixelLowPtUtilities/test/pcsfVerify.ipynb (hunk 5 above)
Comment (VinInn) [inline RecoPixelVertexing/PixelLowPtUtilities/test/pcsfVerify.ipynb]: the png is generated by jupyter itself. I can try to clear the output. it defeats the self documenting purpose though

**Comment by slava77** [comment]
On 7/21/17 10:58 PM, Vincenzo Innocente wrote:
> the png is generated by jupyter itself. I can try to clear the output.
> it defeats the self documenting purpose though
> 

fine for this case.
the image is not that large.

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by VinInn** [comment]
The customizer for HLT went lost! in some rebase????

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
@VinInn 
there is a merge conflict here now

I should catch up with review of this today and hopefully there is no more divergence

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
@Martin-Grunewald  do not understand. what are we supposed to run for 2016 MC  (or reHLT on 2016 data?)

**Comment by Martin-Grunewald** [comment]
The HLT Fake2 menu which is completely stripped down.

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
for the record: the cluster shape filter is significantly less efficient in Phase0 for BPIX1 at least at high lumi
here beginning of run 283408 blue compared to 299592 (black) and 297620 (orange)

image

**Comment by slava77** [comment]
On 7/26/17 2:11 AM, Vincenzo Innocente wrote:
> for the record: the cluster shape filter is significantly less efficient
> in Phase0 for BPIX1 at least at high lumi
> here beginning of run 283408 blue compared to 299592 (black) and 297620
> (orange)

Is this for a "what if we run this on phase0" scenario?
The last version of this PR I looked at did not change phase-0 behavior.

**Comment by VinInn** [comment]
did not change phase-0. Just a side comment...

**Comment by kpedro88** [comment]
assign upgrade

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19824	2004b4b
- changed pixel cluster shape parameter file for filtering (split to two files for phase-1); in line with the description. The MTV plots provided for phase-1 show some improvements in the initialStep (higher eff in the forward at almost constant fake rate); general track efficiency is almost unchanged while the fake rate is up by up to about 10% around |eta| of 1.
- jenkins tests pass and comparisons with baseline show small differences in phase-{1,2} workflows
- local tests roughly confirm the expected behavior with mostly small changes
    - changes in phase-2 are somewhat negligible
    - changes in phase-1 are in line with the MTV plots provided with the PR description (almost unchanged eff, a bit higher fakerate)
- looking at data JetHT 299042 LS41 processed with the promptlike GT (and a snapshot override) in 930pre2, there are some apparent improvements, more significant than what's seen in MC
    - 2% more generalTracks; mostly in the initialStep
    - some increase in globalMuonTracks

all_sign943vsorig_jetht299042c_recotracks_generaltracks__rereco_obj_algo

small increase in bpix1 efficiency
wfjetht299042_hitpatt_eff_global

KShort mass points to an increase in efficiency as well (increase is mostly in the peak; about a 2/3 of the increment has vertex rho below 2 cm still in the pixel acceptance)
all_sign943vsorig_jetht299042c_recovertexcompositecandidates_generalv0candidates_kshort_rereco_obj_mass

**Comment by kpedro88** [comment]
+1

**Comment by slava77** [comment]
@vazzolini, @kmaeshima, @dmitrijus, @vanbesien, @Martin-Grunewald 
DQM and HLT signatures are still missing for this PR.
Please check and propose changes or signoff. This is important for pre3.
Thank you.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by davidlange6** [comment]
merge

**Comment by Dr15Jones** [comment]
I believe this pull request has broken 40+ RelVal workflows in last nights IB.  E.g. 
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_amd64_gcc630/CMSSW_9_3_X_2017-07-28-1100/pyRelValMatrixLogs/run/136.781_RunDoubleEG2017B+RunDoubleEG2017B+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017/step2_RunDoubleEG2017B+RunDoubleEG2017B+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017.log

With 
```
----- Begin Fatal Exception 28-Jul-2017 13:40:08 CEST-----------------------
An exception of category 'FileInPathError' occurred while
   [0] Processing  Event run: 297227 lumi: 33 event: 20650756 stream: 2
   [1] Running path 'HLT_MediumChargedIsoPFTau180HighPtRelaxedIso_Trk50_eta2p1_v5'
   [2] Calling method for module CAHitQuadrupletEDProducer/'hltPixelTracksHitQuadruplets'
   [3] Using EventSetup component ClusterShapeHitFilterESProducer/'' to make data ClusterShapeHitFilter/'ClusterShapeHitFilter' in record CkfComponentsRecord
Exception Message:
edm::FileInPath unable to find file RecoPixelVertexing/PixelLowPtUtilities/data/pixelShape_Phase1TkNewFPix.par anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is: /build/cmsbld/jenkins-workarea/workspace/ib-run-relvals/CMSSW_9_3_X_2017-07-28-1100/src:/build/cmsbld/jenkins-workarea/workspace/ib-run-relvals/CMSSW_9_3_X_2017-07-28-1100/external/slc7_amd64_gcc630/data:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc630/cms/cmssw-patch/CMSSW_9_3_X_2017-07-28-1100/src:/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc630/cms/cmssw-patch/CMSSW_9_3_X_2017-07-28-1100/external/slc7_amd64_gcc630/data
Current directory is: /build/cmsbld/jenkins-workarea/workspace/ib-run-relvals/CMSSW_9_3_X_2017-07-28-1100/pyRelval/136.781_RunDoubleEG2017B+RunDoubleEG2017B+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017
----- End Fatal Exception -------------------------------------------------
```

**Comment by VinInn** [comment]
make sense: still how it passed validation?

**Comment by VinInn** [comment]
no poison data!

**Comment by VinInn** [comment]
#19824 fixit

**Comment by Dr15Jones** [comment]
thanks!

**Pull request merged**